### PR TITLE
Check memcached get resultCode more closely

### PR DESCRIPTION
**Pull request by phil-davis**
Issue #48

### CODE REVIEW

**Review by staabm — APPROVED** [review]
lgtm. a test would be great.

**Comment by phil-davis** [comment]
Test added. I added the test to only MemcachedTest because that implementation does prohibit white-space in the key.

`lib/Memory.php` also implements a "memcache" but it does not prohibit white-space in the key. So the test case will not cause the "expected" exception on that implementation. IMO this does not matter - Memory cache is not meant for real production, and anyway, if someone does try to store keys containing white-space, then they will happily be be set and get without breaking.

File: lib/Memcached.php [hunk 1]
```diff
@@ -47,8 +47,21 @@ public function get($key, $default = null)
...
-        if (false === $result && \Memcached::RES_NOTFOUND === $this->memcached->getResultCode()) {
-            return $default;
+        if (false === $result) {
+            // Note: result can be false because the key exists and has the value boolean false,
+            //       or because something went wrong (cache miss, memcached server timeout...)
+            // So we need to check the result code to work out what to do.
+            $resultCode = $this->memcached->getResultCode();
+            if (\Memcached::RES_BAD_KEY_PROVIDED === $resultCode) {
+                // The key was a string but is invalid. Maybe it was too long, contained whitespace etc.
+                throw new InvalidArgumentException('$key was not valid');
```
Comment (staabm) [inline lib/Memcached.php:57]: this looks like a `\InvalidArgumentException` but not a `\Psr\SimpleCache\InvalidArgumentException::class`

Comment (phil-davis) [reply lib/Memcached.php:57]: This and line 47 are `Sabre\Cache\InvalidArgumentException` and that is:
```
class InvalidArgumentException extends \InvalidArgumentException implements \Psr\SimpleCache\InvalidArgumentException
{
}
```

So it is both `\InvalidArgumentException` and `\Psr\SimpleCache\InvalidArgumentException`

That is existing code. Does that seem correct/reasonable?

**Pull request merged**